### PR TITLE
Osman Elsahib-Database-Week-2

### PR DESCRIPTION
**Pull request by elsahib**
Hello Mentors,

Please review my work for week 2.

Thanks,

### CODE REVIEW

File: week-2/mandatory/2-ecommerce-db/task.md [hunk 1]
```diff
@@ -8,7 +8,20 @@ Below you will find a set of tasks for you to complete to set up a database for
 
 To submit this homework write the correct commands for each question here:
 ```sql
-
+1. select * from customers where country = 'United States';
+2. select * from customers order by name;
+3. select * from products where product_name like '%socks%';
+4. select a.prod_id, p.product_name,a.supp_id,a.unit_price from product_availability a join products p on p.id = a.prod_id where a.unit_price > 100;
+5. select * from product_availability order by unit_price DESC limit 5;
```
Comment (KarenPudner) [inline week-2/mandatory/2-ecommerce-db/task.md:17]: You have correctly identified the 5 most expensive products. Could you do a join here to also include the product name?

File: week-2/mandatory/2-ecommerce-db/task.md [hunk 2]
```diff
@@ -8,7 +8,20 @@ Below you will find a set of tasks for you to complete to set up a database for
...
+6. select p.product_name,a.unit_price,s.supplier_name from product_availability a join products p on p.id = a.prod_id join suppliers s on a.supp_id = s.id;
+7. select p.product_name,s.supplier_name from product_availability a join products p on p.id = a.prod_id join suppliers s on a.supp_id = s.id where s.country = 'United Kingdom';
+8-a. select o.id as Order_ID,o.order_reference, o.order_date, a.unit_price * i.quantity as total from order_items i join orders o on o.id=i.order_id join product_availability a on (i.product_id = a.prod_id and i.supplier_id = a.supp_id) where o.customer_id = 1;
+8-b. select o.id as Order_ID,p.product_name, o.order_reference, o.order_date, a.unit_price * i.quantity as total from order_items i join orders o on o.id=i.order_id join product_availability a on (i.product_id = a.prod_id and i.supplier_id = a.supp_id) join products p on p.id = a.prod_id where o.customer_id = 1;
```
Comment (KarenPudner) [inline week-2/mandatory/2-ecommerce-db/task.md:21]: Some of your results rows have identical order ID, order reference and date, so they are actually the same order. Could you use GROUP BY to aggregate your results by order?

File: week-2/mandatory/3-api/server.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        res.json(result.rows);
+      });
+});
+
+app.get("/products", (req, res) => {
+    db.query("SELECT p.product_name, a.unit_price , s.supplier_name from products p join product_availability a on p.id = a.prod_id join suppliers s on s.id = a.supp_id", (error, result) => {
+        res.json(result.rows);
+      });
+});
+
```
Comment (KarenPudner) [inline week-2/mandatory/3-api/server.js:379]: Ideally handle the error case in each function - there can sometimes be problems with database connectivity, so you should return a suitable error message in this case. Also, return an appropriate status code for the successful responses and the error responses

**Review by KarenPudner — APPROVED** [review]
Homework for week 2. See the comments for a few small improvements you could make, but overall this is excellent work.

File: week-2/mandatory/2-ecommerce-db/task.md (hunk 1 above)
Comment (elsahib) [reply week-2/mandatory/2-ecommerce-db/task.md:17]: `select a.supp_id, a.unit_price, p.product_name from product_availability a join products p on p.id = a.prod_id order by a.unit_price DESC limit 5;`

```
 supp_id | unit_price |  product_name
---------+------------+-----------------
       1 |        299 | Mobile Phone X
       4 |        249 | Mobile Phone X
       2 |         41 | Javascript Book
       1 |         40 | Javascript Book
       3 |         39 | Javascript Book
(5 rows)

```
is this what you mean?

File: week-2/mandatory/2-ecommerce-db/task.md (hunk 2 above)
Comment (elsahib) [reply week-2/mandatory/2-ecommerce-db/task.md:21]: I didn't quite understand what you mean, but these are returning distict results, it's listing all the items of each order!
for 8-b here are the results:
```
 order_id |      product_name       | order_reference | order_date | total
----------+-------------------------+-----------------+------------+-------
        1 | Tee Shirt Olympic Games | ORD001          | 2019-06-01 |    18
        1 | Super warm socks        | ORD001          | 2019-06-01 |    25
        2 | Super warm socks        | ORD002          | 2019-07-15 |    32
        2 | Le Petit Prince         | ORD002          | 2019-07-15 |    10
        3 | Coffee Cup              | ORD003          | 2019-07-11 |    40
        3 | Ball                    | ORD003          | 2019-07-11 |    40
(6 rows)
```

and here are the results for 8-a:
```
 order_id | order_reference | order_date | total
----------+-----------------+------------+-------
        1 | ORD001          | 2019-06-01 |    18
        1 | ORD001          | 2019-06-01 |    25
        2 | ORD002          | 2019-07-15 |    32
        2 | ORD002          | 2019-07-15 |    10
        3 | ORD003          | 2019-07-11 |    40
        3 | ORD003          | 2019-07-11 |    40
(6 rows)

```
and in here I added the product name to the same query in 8-a:
```
 order_id | order_reference |      product_name       | order_date | total
----------+-----------------+-------------------------+------------+-------
        1 | ORD001          | Tee Shirt Olympic Games | 2019-06-01 |    18
        1 | ORD001          | Super warm socks        | 2019-06-01 |    25
        2 | ORD002          | Super warm socks        | 2019-07-15 |    32
        2 | ORD002          | Le Petit Prince         | 2019-07-15 |    10
        3 | ORD003          | Coffee Cup              | 2019-07-11 |    40
        3 | ORD003          | Ball                    | 2019-07-11 |    40
(6 rows)
```
Am I missing something?

File: week-2/mandatory/3-api/server.js (hunk 3 above)
Comment (elsahib) [reply week-2/mandatory/3-api/server.js:379]: I made some basic checking and pushed it github if you can review it.

Comment (KarenPudner) [reply week-2/mandatory/3-api/server.js:379]: Yes, that's exactly right.

File: week-2/mandatory/2-ecommerce-db/task.md (hunk 2 above)
Comment (KarenPudner) [reply week-2/mandatory/2-ecommerce-db/task.md:21]: I meant to aggregate the rows which have the same order ID, order reference and order date, so you would have only 3 rows - the first row would be 1,ORD001, 2019-06-01, 43. (18 + 25 = 43). You can do that by using GROUP BY.

Comment (elsahib) [reply week-2/mandatory/2-ecommerce-db/task.md:21]: Hi Karen, 
I was misled by the question as it was asking for order items to be retrieved as well. here Is what I think you mean.

```sql
select o.id as Order_ID, o.order_reference, o.order_date, sum(a.unit_price * i.quantity) as total from order_items i join orders o on o.id=i.order_id join product_availability a on (i.product_id = a.prod_id and i.supplier_id = a.supp_id) where o.customer_id = 1 group by o.id;
```

```
 order_id | order_reference | order_date | total
----------+-----------------+------------+-------
        1 | ORD001          | 2019-06-01 |    43
        2 | ORD002          | 2019-07-15 |    42
        3 | ORD003          | 2019-07-11 |    80
(3 rows)
```

I've pushed that as 8-c

Comment (KarenPudner) [reply week-2/mandatory/2-ecommerce-db/task.md:21]: That's great. Actually, you're right, the question does ask for order items. The example answers that the volunteers have been given does the GROUP BY, which is why I asked for it, but you answered the question correctly the first time - sorry, should have read the question more closely! But anyway, it's good to practice GROUP BY and you've done it exactly right/

**Comment by elsahib** [comment]
Thanks for your time Karen.

**Pull request closed without merging**